doc/tutorials/namespaces.md

```diff
@@ -1,27 +1,93 @@
-# Namespaces #
+# Understanding Namespaces #
 
-## Definition ##
+## Structure of the key database ##
 
-In Elektra multiple keys may refer to the same part of the configuration.
-A _namespace_ is the part of the key that refers to the key's origin (like a unique path).
+The _key database_ of Elektra is _hierarchically structured_. This means that keys are organized similar to directories in a file system.
 
-Possible sources are: 
+Lets add some keys to the database. To add a key we can use this command:
 
-- **spec/something** for specification of other keys
-- **proc/something** for in-memory keys (e.g. command-line)
-- **dir/something** for dir keys in current working directory
-- **system/something** for system keys in /etc or /
-- **user/something** for user keys in home directory
-- **/something** for cascading keys (actually refers to one of the above)
+	kdb set <key> <value>
 
-Having namespaces enables both admins and users to set specific parts of the application's configuration.
+Now add the the key **/a** with the Value **Value 1** and the key **/b/c** with the Value **Value 2**:
+
+	kdb set /a 'Value 1'
+	kdb set /b/c 'Value 2'
+
+![alt text](img/elektra_hierarchy.svg "Hierarchical structure of key database")
+
+Here you see the internal structure of the database after adding the keys **/a** and **/b/c**.
+For instance the key **/b/c** has the path **/** -> **b** -> **c**.
+
+Note how the name of the key determines the path to its value.
+
+You can use the file system analogy as a mnemonic to remember these commands (like the file system commands in your favorite operating system):
+- `kdb ls <path>`
+	lists keys below _path_
+- `kdb rm <key>`
+	removes a _key_
+- `kdb cp <source> <dest>`
+	copies a key to another path
+- `kdb get <key>`
+	gets the value of _key_
+
+For example `kdb get /b/c` should return `Value 2` now, if you set the values before.
+
+## Namespaces ##
+
+Now we abandon the file system analogy and introduce the concept of _namespaces_.
+
+Every key in Elektra belongs to one of these namespaces:
+- **spec** for specification of other keys
+- **proc** for in-memory keys (e.g. command-line)
```

Comment:
It cannot be used via `kdb` but only within applications. Applications can use this namespace to have some place to override their configuration from some in-memory information they got (e.g. transient user decisions, parsing of command-line information). For example command-line arguments could be handled the following way (pseudo code):

```
ksLookup (ks, "/hello", 0); // returns something from the configuration files
if (!strcmp(argv[n], "--hello")) ksAppendKey (ks, keyNew ("proc/hello", KEY_VALUE, "info from commandline", KEY_END));
ksLookup (ks, "/hello", 0); // now the info from commandline wins.
```